File: tasks/rezoning_geocode_external_merge/code/convert_chicago_bulkgeo_xlsx.py
import argparse
import hashlib
import json
import math
import re
from datetime import datetime, timezone
from pathlib import Path

import pandas as pd

POINT_RE = re.compile(r"POINT\s*\(\s*([-0-9.]+)\s+([-0-9.]+)\s*\)", re.IGNORECASE)
PAIR_RE = re.compile(r"\(?\s*([-0-9.]+)\s*[,;]\s*([-0-9.]+)\s*\)?")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--in-xlsx", required=True)
    parser.add_argument("--out-csv", required=True)
    parser.add_argument("--out-meta-json")
    return parser.parse_args()


def ensure_parent(path: str) -> None:
    Path(path).parent.mkdir(parents=True, exist_ok=True)


def choose_column(columns: list[str], options: list[str], contains: list[str] | None = None) -> str | None:
    lower_to_original = {col.lower(): col for col in columns}
    for option in options:
        if option.lower() in lower_to_original:
            return lower_to_original[option.lower()]
    if contains:
        for col in columns:
            c = col.lower()
            if all(fragment in c for fragment in contains):
                return col
    return None


def safe_float(value) -> float | None:
    if pd.isna(value):
        return None
    try:
        out = float(str(value).strip())
    except Exception:  # noqa: BLE001
        return None
    if math.isnan(out) or math.isinf(out):
        return None
    return out


def parse_lat_lon_text(value: str | None) -> tuple[float | None, float | None]:
    if value is None or pd.isna(value):
        return None, None
    text = str(value).strip()
    if not text or text in {"---", "nan", "None"}:
        return None, None

    point = POINT_RE.search(text)
    if point:
        lon = safe_float(point.group(1))
        lat = safe_float(point.group(2))
        return lat, lon

    pair = PAIR_RE.search(text)
    if pair:
        first = safe_float(pair.group(1))
        second = safe_float(pair.group(2))
        if first is None or second is None:
            return None, None
        if -90 <= first <= 90 and -180 <= second <= 180:
            return first, second
        if -90 <= second <= 90 and -180 <= first <= 180:
            return second, first

    return None, None


def file_sha256(path: str) -> str:
    digest = hashlib.sha256()
    with open(path, "rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def main() -> int:
    args = parse_args()

    df = pd.read_excel(args.in_xlsx, dtype=str)
    cols = df.columns.tolist()

    row_col = choose_column(cols, ["ROW_NUMBER", "row_number", "rownum"], contains=["row", "number"])
    address_col = choose_column(cols, ["address", "input_address"], contains=["address"])
    full_address_col = choose_column(cols, ["full address", "full_address", "matched_address"], contains=["full", "address"])
    status_col = choose_column(cols, ["status", "match_status"], contains=["status"])
    lat_col = choose_column(cols, ["latitude", "lat"], contains=["lat"])
    lon_col = choose_column(cols, ["longitude", "lon", "lng"], contains=["lon"])
    latlon_col = choose_column(cols, ["lat/long coordinates", "coordinates", "location", "point"], contains=["coord"])

    out = pd.DataFrame()
    if row_col:
        out["upload_row_number"] = pd.to_numeric(df[row_col], errors="coerce").astype("Int64")
    else:
        out["upload_row_number"] = range(1, len(df) + 1)

    out["address"] = df[address_col].astype(str) if address_col else None
    out["matched_address"] = df[full_address_col].astype(str) if full_address_col else None
    out["status"] = df[status_col].astype(str) if status_col else None

    out["latitude"] = df[lat_col].map(safe_float) if lat_col else None
    out["longitude"] = df[lon_col].map(safe_float) if lon_col else None

    if latlon_col:
        parsed = df[latlon_col].map(parse_lat_lon_text)
        parsed_lat = parsed.map(lambda p: p[0])
        parsed_lon = parsed.map(lambda p: p[1])
        out["latitude"] = out["latitude"].where(out["latitude"].notna(), parsed_lat)
        out["longitude"] = out["longitude"].where(out["longitude"].notna(), parsed_lon)
        out["lat_long_coordinates"] = df[latlon_col].astype(str)
    else:
        out["lat_long_coordinates"] = None

    out = out.sort_values("upload_row_number", kind="stable")

    ensure_parent(args.out_csv)
    out.to_csv(args.out_csv, index=False)

    if args.out_meta_json:
        ensure_parent(args.out_meta_json)
        summary = {
            "timestamp_utc": datetime.now(timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z"),
            "source_xlsx": args.in_xlsx,
            "source_xlsx_sha256": file_sha256(args.in_xlsx),
            "rows_in_xlsx": int(len(df)),
            "rows_exported": int(len(out)),
            "rows_with_valid_coords": int((out["latitude"].between(-90, 90) & out["longitude"].between(-180, 180)).sum()),
            "detected_columns": {
                "row_col": row_col,
                "address_col": address_col,
                "full_address_col": full_address_col,
                "status_col": status_col,
                "lat_col": lat_col,
                "lon_col": lon_col,
                "latlon_col": latlon_col,
            },
            "output_csv": args.out_csv,
        }
        with open(args.out_meta_json, "w", encoding="utf-8") as handle:
            json.dump(summary, handle, indent=2)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tasks/rezoning_geocode_external_merge/code/convert_chicago_bulkgeo_xlsx.R
## run this line when editing code in Rstudio
# setwd("/Users/jacobherbstman/Desktop/alderman_data/tasks/rezoning_geocode_external_merge/code")
#
## interactive argument examples (mirror Makefile inputs)
# convert_chicago_bulkgeo_xlsx.R --date-tag 20101201_20260212

source("../../setup_environment/code/packages.R")

parse_args <- function(args) {
  out <- list(date_tag = "20101201_20260212")
  i <- 1
  while (i <= length(args)) {
    arg <- args[[i]]
    if (startsWith(arg, "--date-tag=")) {
      out$date_tag <- sub("^--date-tag=", "", arg)
    } else if (arg == "--date-tag" && i < length(args)) {
      i <- i + 1
      out$date_tag <- args[[i]]
    }
    i <- i + 1
  }
  out
}

choose_column <- function(columns, options, contains = NULL) {
  lower <- tolower(columns)
  for (opt in options) {
    idx <- match(tolower(opt), lower)
    if (!is.na(idx)) return(columns[[idx]])
  }
  if (!is.null(contains)) {
    for (col in columns) {
      c <- tolower(col)
      if (all(vapply(contains, function(fragment) stringr::str_detect(c, stringr::fixed(fragment)), logical(1)))) {
        return(col)
      }
    }
  }
  NULL
}

safe_float <- function(value) {
  if (is.na(value)) return(NA_real_)
  out <- suppressWarnings(as.numeric(trimws(as.character(value))))
  if (is.na(out) || is.infinite(out)) return(NA_real_)
  out
}

parse_lat_lon_text <- function(value) {
  if (is.na(value) || trimws(as.character(value)) == "") {
    return(c(NA_real_, NA_real_))
  }
  text <- trimws(as.character(value))
  if (text %in% c("---", "nan", "None")) {
    return(c(NA_real_, NA_real_))
  }

  point_match <- stringr::str_match(text, "POINT\\s*\\(\\s*([-0-9.]+)\\s+([-0-9.]+)\\s*\\)")
  if (!is.na(point_match[[1, 2]])) {
    lon <- safe_float(point_match[[1, 2]])
    lat <- safe_float(point_match[[1, 3]])
    return(c(lat, lon))
  }

  pair_match <- stringr::str_match(text, "\\(?\\s*([-0-9.]+)\\s*[,;]\\s*([-0-9.]+)\\s*\\)?")
  if (!is.na(pair_match[[1, 2]])) {
    first <- safe_float(pair_match[[1, 2]])
    second <- safe_float(pair_match[[1, 3]])
    if (is.na(first) || is.na(second)) return(c(NA_real_, NA_real_))
    if (first >= -90 && first <= 90 && second >= -180 && second <= 180) return(c(first, second))
    if (second >= -90 && second <= 90 && first >= -180 && first <= 180) return(c(second, first))
  }

  c(NA_real_, NA_real_)
}

file_sha256 <- function(path) {
  out <- suppressWarnings(system2("shasum", c("-a", "256", path), stdout = TRUE, stderr = FALSE))
  if (length(out) == 0) return(NA_character_)
  trimws(strsplit(out[[1]], " ", fixed = TRUE)[[1]][[1]])
}

main <- function() {
  opts <- parse_args(commandArgs(trailingOnly = TRUE))
  in_xlsx <- paste0("../input/chicago_geocoder_results_", opts$date_tag, ".xlsx")
  out_csv <- paste0("../input/chicago_geocoder_results_", opts$date_tag, ".csv")
  out_meta_json <- paste0("../input/chicago_geocoder_results_", opts$date_tag, "_import_meta.json")

  df <- readxl::read_xlsx(in_xlsx, col_types = "text")
  cols <- names(df)

  row_col <- choose_column(cols, c("ROW_NUMBER", "row_number", "rownum"), contains = c("row", "number"))
  address_col <- choose_column(cols, c("address", "input_address"), contains = c("address"))
  full_address_col <- choose_column(cols, c("full address", "full_address", "matched_address"), contains = c("full", "address"))
  status_col <- choose_column(cols, c("status", "match_status"), contains = c("status"))
  lat_col <- choose_column(cols, c("latitude", "lat"), contains = c("lat"))
  lon_col <- choose_column(cols, c("longitude", "lon", "lng"), contains = c("lon"))
  latlon_col <- choose_column(cols, c("lat/long coordinates", "coordinates", "location", "point"), contains = c("coord"))

  out <- tibble::tibble()
  if (!is.null(row_col)) {
    out$upload_row_number <- suppressWarnings(as.integer(df[[row_col]]))
  } else {
    out$upload_row_number <- seq_len(nrow(df))
  }

  out$address <- if (!is.null(address_col)) as.character(df[[address_col]]) else NA_character_
  out$matched_address <- if (!is.null(full_address_col)) as.character(df[[full_address_col]]) else NA_character_
  out$status <- if (!is.null(status_col)) as.character(df[[status_col]]) else NA_character_

  out$latitude <- if (!is.null(lat_col)) vapply(df[[lat_col]], safe_float, numeric(1)) else NA_real_
  out$longitude <- if (!is.null(lon_col)) vapply(df[[lon_col]], safe_float, numeric(1)) else NA_real_

  if (!is.null(latlon_col)) {
    parsed <- t(vapply(df[[latlon_col]], parse_lat_lon_text, numeric(2)))
    parsed_lat <- parsed[, 1]
    parsed_lon <- parsed[, 2]
    out$latitude[is.na(out$latitude)] <- parsed_lat[is.na(out$latitude)]
    out$longitude[is.na(out$longitude)] <- parsed_lon[is.na(out$longitude)]
    out$lat_long_coordinates <- as.character(df[[latlon_col]])
  } else {
    out$lat_long_coordinates <- NA_character_
  }

  out <- out %>%
    dplyr::arrange(.data$upload_row_number)

  dir.create("../input", recursive = TRUE, showWarnings = FALSE)
  readr::write_csv(out, out_csv)

  rows_with_valid_coords <- sum(
    !is.na(out$latitude) &
      !is.na(out$longitude) &
      out$latitude >= -90 & out$latitude <= 90 &
      out$longitude >= -180 & out$longitude <= 180
  )

  summary <- list(
    timestamp_utc = format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC"),
    source_xlsx = in_xlsx,
    source_xlsx_sha256 = file_sha256(in_xlsx),
    rows_in_xlsx = nrow(df),
    rows_exported = nrow(out),
    rows_with_valid_coords = as.integer(rows_with_valid_coords),
    detected_columns = list(
      row_col = row_col,
      address_col = address_col,
      full_address_col = full_address_col,
      status_col = status_col,
      lat_col = lat_col,
      lon_col = lon_col,
      latlon_col = latlon_col
    ),
    output_csv = out_csv
  )

  jsonlite::write_json(summary, out_meta_json, pretty = TRUE, auto_unbox = TRUE, na = "null")
}

main()
